iped-carvers/iped-carvers-impl/src/main/java/iped/carvers/custom/TorrentCarver.java

```diff
@@ -0,0 +1,62 @@
+package iped.carvers.custom;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+
+import com.dampcake.bencode.BencodeInputStream;
+
+import iped.carvers.api.Hit;
+import iped.carvers.standard.DefaultCarver;
+import iped.data.IItem;
+import iped.io.SeekableInputStream;
+import iped.utils.IOUtil;
+
+public class TorrentCarver extends DefaultCarver {
+    @Override
+    public boolean isSpecificIgnoreCorrupted() {
+        return true;
+    }
+
+    @Override
+    public long getLengthFromHit(IItem parentEvidence, Hit header) throws IOException {
+        SeekableInputStream is = null;
+        BencodeInputStream bis = null;
+        try {
+            long off = header.getOffset();
+            is = parentEvidence.getSeekableInputStream();
+            is.seek(header.getOffset());
+            bis = new BencodeInputStream(is, StandardCharsets.UTF_8, true);
+            Map<String, Object> dict = bis.readDictionary();
```

Comment:
readDictionary will throw exception if incomplete dictionary is found. Not sure if it is the desired, maybe some info can still be valuable, even if an incomplete dict is found.
Maybe the carve can occur, although the parse will fail.

Reply:
That was intentional, to avoid bringing too many invalid (and useless) items. 
Of course there is always a choice between losing some information and bringing too much garbage.

Reply:
Sure. But I'm also affraid with some false negative. Maybe some relevant incomplete piece can be ignored. The method was designed exactly to avoid too many false positives. Maybe the name "isValid" is to "direct" as the idea was to exclude totally incomplete items.
For instance the following piece will be ignored as it is incomplete:
`00041eda  64 34 3a 69 6e 66 6f 64  35 3a 66 69 6c 65 73 6c  |d4:infod5:filesl|
00041eea  64 36 3a 6c 65 6e 67 74  68 69 34 33 34 36 33 39  |d6:lengthi434639|
00041efa  34 35 31 31 65 34 3a 70  61 74 68 6c 39 39 3a 41  |4511e4:pathl99:A|
00041f0a  76 65 73 2e 64 65 2e 52  61 70 69 6e 61 2e 41 72  |ves.de.Rapina.Ar|
00041f1a  6c 65 71 75 69 6e 61 2e  65 2e 73 75 61 2e 45 6d  |lequina.e.sua.Em|
00041f2a  61 6e 63 69 70 61 63 61  6f 2e 46 61 6e 74 61 62  |ancipacao.Fantab|
00041f3a  75 6c 6f 73 61 2e 32 30  32 30 2e 31 30 38 30 70  |ulosa.2020.1080p|
00041f4a  2e 57 45 42 2d 44 4c 2e  44 44 35 2e 31 2e 48 32  |.WEB-DL.DD5.1.H2|
00041f5a  36 34 2d 45 56 4f 2e 44  55 41 4c 2d 52 4b 2e 6d  |64-EVO.DUAL-RK.m|
00041f6a  6b 76 65 65 64 36 3a 6c  65 6e 67 74 68 69 33 30  |kveed6:lengthi30|
00041f7a  37 65 34 3a 70 61 74 68  6c 31 30 36 3a 41 76 65  |7e4:pathl106:Ave|
00041f8a  73 2e 64 65 2e 52 61 70  69 6e 61 2e 41 72 6c 65  |s.de.Rapina.Arle|
00041f9a  71 75 69 6e 61 2e 65 2e  73 75 61 2e 45 6d 61 6e  |quina.e.sua.Eman|
00041faa  63 69 70 61 63 61 6f 2e  46 61 6e 74 61 62 75 6c  |cipacao.Fantabul|
00041fba  6f 73 61 2e 32 30 32 30  2e 31 30 38 30 70 2e 57  |osa.2020.1080p.W|
00041fca  45 42 2d 44 4c 2e 44 44  35 2e 31 2e 48 32 36 34  |EB-DL.DD5.1.H264|
00041fda  2d 45 56 4f 2e 44 55 41  4c 2d 52 4b 5f 46 6f 72  |-EVO.DUAL-RK_For|
00041fea  63 65 64 2e 73 72 74 65  65 64 36 3a 6c 65 6e 67  |ced.srteed6:leng|
00041ffa  74 68 69 31 30 32 30 33  31 65 34 3a 70 61 74 68  |thi102031e4:path|
0004200a  6c 37 3a 4c 65 67 65 6e  64 61 39 39 3a 41 76 65  |l7:Legenda99:Ave|
0004201a  73 2e 64 65 2e 52 61 70  69 6e 61 2e 41 72 6c 65  |s.de.Rapina.Arle|
0004202a  71 75 69 6e 61 2e 65 2e  73 75 61 2e 45 6d 61 6e  |quina.e.sua.Eman|
0004203a  63 69 70 61 63 61 6f 2e  46 61 6e 74 61 62 75 6c  |cipacao.Fantabul|
0004204a  6f 73 61 2e 32 30 32 30  2e 31 30 38 30 70 2e 57  |osa.2020.1080p.W|
0004205a  45 42 2d 44 4c 2e 44 44  35 2e 31 2e 48 32 36 34  |EB-DL.DD5.1.H264|
0004206a  2d 45 56 4f 2e 44 55 41  4c 2d 52 4b 2e 73 72 74  |-EVO.DUAL-RK.srt|
0004207a  65 65 65 34 3a 6e 61 6d  65 39 35 3a 41 76 65 73  |eee4:name95:Aves|
0004208a  2e 64 65 2e 52 61 70 69  6e 61 2e 41 72 6c 65 71  |.de.Rapina.Arleq|
0004209a  75 69 6e 61 2e 65 2e 73  75 61 2e 45 6d 61 6e 63  |uina.e.sua.Emanc|
000420aa  69 70 61 63 61 6f 2e 46  61 6e 74 61 62 75 6c 6f  |ipacao.Fantabulo|
000420ba  73 61 2e 32 30 32 30 2e  31 30 38 30 70 2e 57 45  |sa.2020.1080p.WE|
000420ca  42 2d 44 4c 2e 44 44 35  2e 31 2e 48 32 36 34 2d  |B-DL.DD5.1.H264-|
000420da  45 56 4f 2e 44 55 41 4c  2d 52 4b 31 32 3a 70 69  |EVO.DUAL-RK12:pi|
000420ea  65 63 65 20 6c 65 6e 67  74 68 69 31 30 34 38 35  |ece lengthi10485|
000420fa  37 36 65 36 3a 70 69 65  63 65 73 38 32 39 32 30  |76e6:pieces82920|

Reply:
Anyway, I will let you decide.

Reply:
I remembered now that there is a parameter in CarverConfig.xml to ignore corrupted files. So, if implemented this way, even if ignoreCorrupted is set to "false", those entries won't be extracted.

Reply:
Ok, it seems fine this way.